sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs

```diff
@@ -550,6 +592,265 @@ private async Task RunBatchReceiveLoopAsync(CancellationToken cancellationToken)
+                                                   ServiceBusReceiveActions receiveActions,
+                                                   ServiceBusMessageActions messageActions,
+                                                   CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.Value.CreateScope(
+                            _isSessionsEnabled ? Constants.ProcessSessionMessagesActivityName : Constants.ProcessMessagesActivityName,
+                            DiagnosticScope.ActivityKind.Consumer,
+                            MessagingDiagnosticOperation.Process);
+
+            scope.SetMessageData(messagesArray);
+
+            scope.Start();
+            FunctionResult result = await _triggerExecutor.TryExecuteAsync(input.GetTriggerFunctionData(), cancellationToken).ConfigureAwait(false);
+            if (result.Exception != null)
+            {
+                scope.Failed(result.Exception);
+            }
+            receiveActions.EndExecutionScope();
+
+            var processedMessages = messagesArray.Concat(receiveActions.Messages.Keys);
+            // Complete batch of messages only if the execution was successful
+            if (_autoCompleteMessages && result.Succeeded)
+            {
+                List<Task> completeTasks = new List<Task>();
+                foreach (ServiceBusReceivedMessage message in processedMessages)
+                {
+                    // skip messages that were settled in the user's function
+                    if (input.MessageActions.SettledMessages.ContainsKey(message))
+                    {
+                        continue;
+                    }
+
+                    // Pass CancellationToken.None to allow autocompletion to finish even when shutting down
+                    completeTasks.Add(receiver.CompleteMessageAsync(message, CancellationToken.None));
+                }
+
+                await Task.WhenAll(completeTasks).ConfigureAwait(false);
+            }
+            else if (!result.Succeeded)
+            {
+                // For failed executions, we abandon the messages regardless of the autoCompleteMessages configuration.
+                // This matches the behavior that happens for single dispatch functions as the processor does the same thing
+                // in the Service Bus SDK.
+
+                List<Task> abandonTasks = new();
+                foreach (ServiceBusReceivedMessage message in processedMessages)
+                {
+                    // skip messages that were settled in the user's function
+                    if (input.MessageActions.SettledMessages.ContainsKey(message))
+                    {
+                        continue;
+                    }
+
+                    // Pass CancellationToken.None to allow abandon to finish even when shutting down
+                    abandonTasks.Add(receiver.AbandonMessageAsync(message, cancellationToken: CancellationToken.None));
```

Comment:
Consider: should an invocation failure mean that we abandon any messages in the cache? If the next invocation succeeds it would be out of order. (Maybe only abandon if it's a session failure?)

Reply:
I don't think we need to - when there is a failure it is up to the user to decide if they want to close the session and start over which they can do with the session actions.